tools/autograd/derivatives.yaml

```diff
@@ -632,6 +632,9 @@
 - name: polygamma(int n, Tensor self) -> Tensor
   self: grad * polygamma(n + 1, self)
 
+- name: polygamma_(Tensor(a!) self, int n) -> Tensor(a!)
```

Comment:
Since signature of function variant is different from the inplace variant, add explicit entry for inplace variant

Reply:
This is pretty weird. Is this because inplace self must always be in first position? (I know we don't have this restriction for regular functions; c.f., `where.self(Tensor condition, Tensor self, Tensor other) -> Tensor`)

Reply:
Actually where has implicit grad and it depends on broadcast and _s_where (actual implementation of where logic) https://github.com/pytorch/pytorch/blob/6f63126b5c39ee35dd301ff957a155d42c7fdd1f/aten/src/ATen/native/native_functions.yaml#L3771-L3792

Just a guess for why we need `polygamma_` formula explicitly is that the signature of `polygamma` and `polygamma_` differ in the position of self argument.
```
polygamma(int n, Tensor self) -> Tensor
polygamma_(Tensor(a!) self, int n) -> Tensor(a!)
```

I think inplace version only works implicitly if the signature is same as method or function variant.
Maybe @albanD can shed more light.

One thing I have verified is that inplace grad doesn't work without this
```
>>> import torch
>>> torch.__version__
'1.7.1+cpu'
>>> x = torch.tensor([1.], dtype=torch.float32, requires_grad=True)
>>> x.polygamma_(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: polygamma(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
```

Reply:
@albanD, would you like to take a look?

Reply:
> One thing I have verified is that inplace grad doesn't work without this

What is the exact setting where this happens? It seems to indicate that you specified CompositeImplicitAutograd in native_functions and it tries to differerentiate the c++ implementation that uses out=.
Using CompositeExplicitAutograd and no formula in derivatives.yaml should force it to use the formula from the non-inplace case. And since all arguments are the same, that should work.

The one problem I see is that the original formula uses `self` (which is the input value, and we handle that properly when we re-use the out of place formula). But after the inplace, when you specify the formula, the original value does not exist anymore so I am not sure if this formula here is correct.

Reply:
> What is the exact setting where this happens? It seems to indicate that you specified CompositeImplicitAutograd in native_functions and it tries to differerentiate the c++ implementation that uses out=.

That is when it was failing (CompositeImplicitAutograd didn't specify it but I think that is the default if unspecified).

> Using CompositeExplicitAutograd and no formula in derivatives.yaml should force it to use the formula from the non-inplace case. And since all arguments are the same, that should work.

Tried the following diff.
```patch
diff --git a/aten/src/ATen/native/native_functions.yaml b/aten/src/ATen/native/native_functions.yaml
index 3784cd6643..2363a538ff 100644
--- a/aten/src/ATen/native/native_functions.yaml
+++ b/aten/src/ATen/native/native_functions.yaml
@@ -4986,11 +4986,6 @@
   dispatch:
     CPU, CUDA: digamma_
 
-- func: polygamma_(Tensor(a!) self, int n) -> Tensor(a!)
-  variants: method
-  dispatch:
-    CompositeExplicitAutograd: polygamma_
-
 - func: renorm_(Tensor(a!) self, Scalar p, int dim, Scalar maxnorm) -> Tensor(a!)
   variants: method
   dispatch:
@@ -5763,6 +5758,11 @@
   dispatch:
     CompositeExplicitAutograd: polygamma
 
+- func: polygamma_(Tensor(a!) self, int n) -> Tensor(a!)
+  variants: method
+  dispatch:
+    CompositeExplicitAutograd: polygamma_
+
 - func: erfinv(Tensor self) -> Tensor
   variants: method, function
   dispatch:
diff --git a/tools/autograd/derivatives.yaml b/tools/autograd/derivatives.yaml
index 4ecc4c72c7..20b59ad436 100644
--- a/tools/autograd/derivatives.yaml
+++ b/tools/autograd/derivatives.yaml
@@ -632,9 +632,6 @@
 - name: polygamma(int n, Tensor self) -> Tensor
   self: grad * polygamma(n + 1, self)
 
-- name: polygamma_(Tensor(a!) self, int n) -> Tensor(a!)
-  self: grad * polygamma(n + 1, self)
-
 - name: log(Tensor self) -> Tensor
   self: grad.div(self.conj())

```

which gives the following error, `RuntimeError: derivative for polygamma_ is not implemented`

<details>

```
============================================================================= FAILURES =============================================================================
______________________________________________ TestGradientsCPU.test_inplace_grad_polygamma_polygamma_n_0_cpu_float64 ______________________________________________
Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_unary_infra_change/test/test_ops.py", line 165, in test_inplace_grad
    self._grad_test_helper(device, dtype, op, self._get_safe_inplace(op.get_inplace()))
  File "/home/kshiteej/Pytorch/pytorch_unary_infra_change/test/test_ops.py", line 136, in _grad_test_helper
    return self._check_helper(device, dtype, op, variant, 'gradcheck')
  File "/home/kshiteej/Pytorch/pytorch_unary_infra_change/test/test_ops.py", line 120, in _check_helper
    self.assertTrue(gradcheck(fn, (sample.input,) + sample.args,
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 584, in gradcheck
    analytical, failed = check_analytical_jacobian_attributes(tupled_inputs, o, nondet_tol, 1.0,
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 202, in check_analytical_jacobian_attributes
    jacobians_rows = compute_analytical_jacobian_rows(vjp_fn, output.clone(), grad_out_scale)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 243, in compute_analytical_jacobian_rows
    grad_inputs = vjp_fn(grad_out_base)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 199, in backward_fn
    return torch.autograd.grad(output, diff_input_list, grad_output,
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/autograd/__init__.py", line 224, in grad
    return Variable._execution_engine.run_backward(
RuntimeError: derivative for polygamma_ is not implemented
____________________________________________ TestGradientsCPU.test_inplace_gradgrad_polygamma_polygamma_n_0_cpu_float64 ____________________________________________
Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_unary_infra_change/test/test_ops.py", line 185, in test_inplace_gradgrad
    self._gradgrad_test_helper(device, dtype, op, self._get_safe_inplace(op.get_inplace()))
  File "/home/kshiteej/Pytorch/pytorch_unary_infra_change/test/test_ops.py", line 139, in _gradgrad_test_helper
    return self._check_helper(device, dtype, op, variant, 'gradgradcheck')
  File "/home/kshiteej/Pytorch/pytorch_unary_infra_change/test/test_ops.py", line 124, in _check_helper
    self.assertTrue(gradgradcheck(fn, (sample.input,) + sample.args,
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 719, in gradgradcheck
    return gradcheck(
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 572, in gradcheck
    func_out = func(*tupled_inputs)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 716, in new_func
    grad_inputs = torch.autograd.grad(outputs, input_args, grad_outputs, create_graph=True)
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/site-packages/torch/autograd/__init__.py", line 224, in grad
    return Variable._execution_engine.run_backward(
RuntimeError: derivative for polygamma_ is not implemented
```
</details>

> The one problem I see is that the original formula uses self (which is the input value, and we handle that properly when we re-use the out of place formula). But after the inplace, when you specify the formula, the original value does not exist anymore so I am not sure if this formula here is correct.

In that case shouldn't the gradcheck and gradgradcheck fail?

Reply:
@albanD Gentle Ping :)

Reply:
Hey sorry.

I guess this code is tripped up by the re-ordering of the args then: https://github.com/pytorch/pytorch/blob/af1a772876ca52ad75f3b33e83c4314db81d6a48/tools/codegen/api/autograd.py#L199-L211

I guess you will need a special formula for it then if you cannot re-order the arguments.